assembly/src/parsers/blocks.rs

```diff
@@ -251,22 +254,36 @@ impl BlockParser {
 // ================================================================================================
 
 pub fn combine_blocks(mut blocks: Vec<CodeBlock>) -> CodeBlock {
-    // merge consecutive Span blocks
+    // merge consecutive Span blocks.
     let mut merged_blocks: Vec<CodeBlock> = Vec::with_capacity(blocks.len());
+    // Keep track of all the consecutive Span blocks and are merged together when
+    // there is a discontinuity.
+    let mut contiguous_spans: Vec<Span> = Vec::new();
+    // flag for last block being a SPAN block or not.
+    let mut is_last_block_span = false;
+
     blocks.drain(0..).for_each(|block| {
         if block.is_span() {
-            if let Some(CodeBlock::Span(last_span)) = merged_blocks.last_mut() {
-                // this is guaranteed to execute because we know that the block is a span
-                if let CodeBlock::Span(span) = block {
-                    last_span.append(span);
+            // this is guaranteed to execute because we know that the block is a span.
+            if let CodeBlock::Span(span) = block {
+                if is_last_block_span {
+                    contiguous_spans.push(span);
+                } else {
+                    contiguous_spans.push(span);
+                    is_last_block_span = true;
                 }
-            } else {
-                merged_blocks.push(block);
             }
```

Comment:
I think we can simplify this here as:
```Rust
if block.is_span() {
    contiguous_spans.push(span);
    is_last_block_span = true;
} else {
    if is_last_block_span {
        is_last_block_span = false;
        merged_blocks.push(combine_spans(&mut contiguous_spans));
    }
    merged_blocks.push(block);
}
```
You'll need to change the type of `contiguous_spans` to `Vec<CodeBlock>` and then handle the unwrapping in `combine_spans()` - but I think it will be cleaner this way.

Reply:
Done!